domain/src/main/java/com/thoughtworks/go/server/service/InstanceFactory.java

```diff
@@ -25,14 +25,19 @@
 import org.springframework.stereotype.Component;
 
 import java.util.List;
+import java.util.Map;
 
 @Component
 public class InstanceFactory {
 
     public Pipeline createPipelineInstance(PipelineConfig pipelineConfig, BuildCause buildCause, SchedulingContext context, String md5, Clock clock) {
         buildCause.assertMaterialsMatch(pipelineConfig.materialConfigs());
         buildCause.assertPipelineConfigAndMaterialRevisionMatch(pipelineConfig);
-        return new Pipeline(CaseInsensitiveString.str(pipelineConfig.name()), pipelineConfig.getLabelTemplate(), buildCause, createStageInstance(pipelineConfig.first(), context, md5, clock));
+
+        Map<CaseInsensitiveString, String> variables = EnvironmentVariables.toEnvironmentVariables(pipelineConfig.getVariables()).insecureVariablesHash();
```

Comment:
Instead of creating this hash, won't it be easier to just pass along the `EnvironmentVariables` object and use that later, and use only the insecure variables from it. Something like:

```
envVars.valueForInsecureEnvironmentVariable("ABC")
```

Reply:
@ibnc I think you should consider this. That temporary hash seems unnecessary and one more thing for someone to understand.

Reply:
@arvindsv If I do this, then the config module will need the domain module as a dependency. Which will create a circular dependency. There's an `EnvironmentVariables` class accessible from the `PipelineLabel` class, but it's deprecated.

Reply:
Wouldn't the standard approach for that be something like this, then?

```diff
diff --git a/config/config-api/src/main/java/com/thoughtworks/go/domain/label/InsecureEnvironmentVariables.java b/config/config-api/src/main/java/com/thoughtworks/go/domain/label/InsecureEnvironmentVariables.java
index a2f1b6fc87..ac6d1f5e79 100644
--- a/config/config-api/src/main/java/com/thoughtworks/go/domain/label/InsecureEnvironmentVariables.java
+++ b/config/config-api/src/main/java/com/thoughtworks/go/domain/label/InsecureEnvironmentVariables.java
@@ -1,4 +1,23 @@
+/*
+ * Copyright 2018 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package com.thoughtworks.go.domain.label;
 
+import com.thoughtworks.go.config.CaseInsensitiveString;
+
 public interface InsecureEnvironmentVariables {
+    String getInsecureEnvironmentVariableOrDefault(CaseInsensitiveString key, String defaultValue);
 }
diff --git a/config/config-api/src/main/java/com/thoughtworks/go/domain/label/PipelineLabel.java b/config/config-api/src/main/java/com/thoughtworks/go/domain/label/PipelineLabel.java
index acf17d332a..1af8acb166 100644
--- a/config/config-api/src/main/java/com/thoughtworks/go/domain/label/PipelineLabel.java
+++ b/config/config-api/src/main/java/com/thoughtworks/go/domain/label/PipelineLabel.java
@@ -26,7 +26,7 @@ import java.util.regex.Pattern;
 
 public class PipelineLabel implements Serializable {
     protected String label;
-    private Map<CaseInsensitiveString, String> envVars;
+    private InsecureEnvironmentVariables envVars;
     public static final String COUNT = "COUNT";
     public static final String ENV_VAR_PREFIX = "env:";
     public static final String COUNT_TEMPLATE = String.format("${%s}", COUNT);
@@ -35,7 +35,7 @@ public class PipelineLabel implements Serializable {
         this.label = labelTemplate;
     }
 
-    public PipelineLabel(String labelTemplate, Map<CaseInsensitiveString, String> envVars) {
+    public PipelineLabel(String labelTemplate, InsecureEnvironmentVariables envVars) {
         this.label = labelTemplate;
         this.envVars = envVars;
     }
@@ -68,7 +68,7 @@ public class PipelineLabel implements Serializable {
             if (envVars == null) {
                 return "\\" + matcher.group(0);
             }
-            revision = envVars.getOrDefault(new CaseInsensitiveString(material.toString().split(":", 2)[1]), "");
+            revision = envVars.getInsecureEnvironmentVariableOrDefault(new CaseInsensitiveString(material.toString().split(":", 2)[1]), "");
         } else {
             if (!materialRevisions.containsKey(material)) {
                 //throw new IllegalStateException("cannot find material '" + material + "'");
@@ -119,7 +119,7 @@ public class PipelineLabel implements Serializable {
         return create(labelTemplate, null);
     }
 
-    public static PipelineLabel create(String labelTemplate, Map<CaseInsensitiveString, String> envVars) {
+    public static PipelineLabel create(String labelTemplate, InsecureEnvironmentVariables envVars) {
         if (StringUtils.isBlank(labelTemplate)) {
             return defaultLabel();
         } else {
diff --git a/domain/src/main/java/com/thoughtworks/go/domain/EnvironmentVariables.java b/domain/src/main/java/com/thoughtworks/go/domain/EnvironmentVariables.java
index 1fc1a068db..24360d39c9 100644
--- a/domain/src/main/java/com/thoughtworks/go/domain/EnvironmentVariables.java
+++ b/domain/src/main/java/com/thoughtworks/go/domain/EnvironmentVariables.java
@@ -19,13 +19,14 @@ package com.thoughtworks.go.domain;
 import com.thoughtworks.go.config.CaseInsensitiveString;
 import com.thoughtworks.go.config.EnvironmentVariableConfig;
 import com.thoughtworks.go.config.EnvironmentVariablesConfig;
+import com.thoughtworks.go.domain.label.InsecureEnvironmentVariables;
 import com.thoughtworks.go.util.command.EnvironmentVariableContext;
 
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
-public class EnvironmentVariables extends BaseCollection<EnvironmentVariable> {
+public class EnvironmentVariables extends BaseCollection<EnvironmentVariable> implements InsecureEnvironmentVariables {
 
     private static final String JOB = EnvironmentVariableType.Job.toString();
 
@@ -73,4 +74,9 @@ public class EnvironmentVariables extends BaseCollection<EnvironmentVariable> {
         }
         return environmentVariables;
     }
+
+    @Override
+    public String getInsecureEnvironmentVariableOrDefault(CaseInsensitiveString key, String defaultValue) {
+        return null;
+    }
 }
diff --git a/domain/src/main/java/com/thoughtworks/go/domain/Pipeline.java b/domain/src/main/java/com/thoughtworks/go/domain/Pipeline.java
index d605f815ae..7848880df6 100644
--- a/domain/src/main/java/com/thoughtworks/go/domain/Pipeline.java
+++ b/domain/src/main/java/com/thoughtworks/go/domain/Pipeline.java
@@ -20,6 +20,7 @@ import com.thoughtworks.go.config.CaseInsensitiveString;
 import com.thoughtworks.go.config.CruiseConfig;
 import com.thoughtworks.go.config.materials.Materials;
 import com.thoughtworks.go.domain.buildcause.BuildCause;
+import com.thoughtworks.go.domain.label.InsecureEnvironmentVariables;
 import com.thoughtworks.go.domain.label.PipelineLabel;
 
 import java.io.File;
@@ -53,7 +54,7 @@ public class Pipeline extends PersistentObject implements PipelineInfo {
         this.pipelineLabel = PipelineLabel.create(labelTemplate);
     }
 
-    public Pipeline(String pipelineName, String labelTemplate, BuildCause buildCause, Map<CaseInsensitiveString, String> envVars, Stage... stages) {
+    public Pipeline(String pipelineName, String labelTemplate, BuildCause buildCause, InsecureEnvironmentVariables envVars, Stage... stages) {
         this(new Stages(stages));
         this.pipelineName = pipelineName;
         this.buildCause = buildCause;
diff --git a/domain/src/main/java/com/thoughtworks/go/server/service/InstanceFactory.java b/domain/src/main/java/com/thoughtworks/go/server/service/InstanceFactory.java
index 090de345b9..6acb05eae1 100644
--- a/domain/src/main/java/com/thoughtworks/go/server/service/InstanceFactory.java
+++ b/domain/src/main/java/com/thoughtworks/go/server/service/InstanceFactory.java
@@ -34,10 +34,7 @@ public class InstanceFactory {
         buildCause.assertMaterialsMatch(pipelineConfig.materialConfigs());
         buildCause.assertPipelineConfigAndMaterialRevisionMatch(pipelineConfig);
 
-        Map<CaseInsensitiveString, String> variables = EnvironmentVariables.toEnvironmentVariables(pipelineConfig.getVariables()).insecureVariablesHash();
-        variables.putAll(buildCause.getVariables().insecureVariablesHash());
-
-        return new Pipeline(CaseInsensitiveString.str(pipelineConfig.name()), pipelineConfig.getLabelTemplate(), buildCause, variables, createStageInstance(pipelineConfig.first(), context, md5, clock));
+        return new Pipeline(CaseInsensitiveString.str(pipelineConfig.name()), pipelineConfig.getLabelTemplate(), buildCause, pipelineConfig.getVariables(), createStageInstance(pipelineConfig.first(), context, md5, clock));
     }
 
     public Stage createStageInstance(PipelineConfig pipelineConfig, CaseInsensitiveString stageName, SchedulingContext context, String md5, Clock clock) {
```